Preserve visual layer of meshes in GridMap nodes

GlyphTheWolf opened the pull request:
### Description
This PR adds visual layer property to `MeshLibrary` items allowing to preserve `MeshInstance` nodes visual layer when used in `GridMap` nodes.

That is long missing feature for `GridMap` node requested even since Godot 3 (see #40245) and possibility for setting different render layers may be required to achieve some rendering effects.

This change should work also for `GridMap`s using baked meshes.

#### Note 1:
To use this new feature you need to recreate your `MeshLibrary` resource (export your tileset source scene as `MeshLibrary` again) or update your old `MeshLibrary` items visual layer by hand.

#### Note 2:
If you are already using baked meshes in your project to use new feature you need to invalidate old baked meshes (for example by modifying `GridMap` cells) and re-bake.

### Example
As an example usage now it is possible to use `Decal` cull property to draw some "selection area" only on floors and avoid texture being mapped also on walls (impossible before when using `GridMap` for floors & walls).

![GridmapRenderLayers](https://github.com/user-attachments/assets/cd2eb0d8-a0a0-44bd-b97f-14b50f869ff8)

GlyphTheWolf committed:
Add render layers to MeshLibrary

smix8 commented:
Hi, thanks for working on this. Can you link to the mentioned discussions / proposal(s) for this feature.

Not opposed to adding this but layers and properties targeted at cell items shouldn't be global for the entire GridMap that is a Node3D all about cell item customization (the physics material on the GridMap node is a legacy mistake).

The better place to add layer bitmasks affecting cell items is inside the MeshLibrary per item, not on the GridMap node directly.
See the MeshLibrary cell item `navigation_layers` or PR https://github.com/godotengine/godot/pull/85443 as an example.

Adding any bitmasks to cell items requires to update the MeshLibrary scene exporter to export all the related layer bitmasks as well.

GlyphTheWolf commented:
Yes, I even have my custom version where I updated it in `MeshLibrary` and it works per item. That should be fine for render layers, but:
1. It will change `MeshLibrary` resource, won't it be breaking compatibility then? Not sure how Godot handle reading old resource when new properties were added.
2. If render layer should be per item shouldn't collision layer/mask/priority also be per item? With only one collision layer per `GridMap` It may be still required to just use multiple `GridMap`s and as I analyzed it seems supporting collisions per item, may be not so easy (possible, but require more changes)
3. In some use cases still need multiple `GridMap`s even if implemented 1 & 2

That's why I prepared that simple version in my request. I can update it to support render layer per item, but just this simple change should give much more control anyway and it is relatively small and simple.

smix8 commented:
Since the rendering layers would be a new property it is not really a compatibility problem as older resourcse would just start with the default.

Physics would be more a compatibility problem to change and the last time it was brought up with contributors there was no real decision on it so it just kept status quo. Right now all those physics properties are on the GridMap node and the GridMap Octants also merge the collision shapes under a single body with this assumption. It would need to create differen bodies for the octants when cell items with different physics properties are used. Same as it creates different multimeshes for each different cell item with a visual mesh in the octant.

GlyphTheWolf commented:
I modified code and moved render layers from `GridMap` node to `MeshLibrary` resource. Now it is possible to have multiple render layers combinations per single `GridMap` (different for each `MeshLibrary` item). When converting scene to `MeshLibrary` render layers from `MeshInstance3D` nodes are saved per item and then same render layer setting is used when placed in `GridMap`.

I also made it work with baked meshes, but probably someone who knows more about that should check it, since I had to keep/pass some additional data to make it working. I hope it won't break anything else.

Here are example from my test with baked lightmap (I also tested without lightmaps and it works fine):

![RenderLayersMeshLib1](https://github.com/user-attachments/assets/f21b37f7-6e08-4df1-b117-db382fdd51e5)
![RenderLayersMeshLib2](https://github.com/user-attachments/assets/b03bd3a8-d4ba-4d92-a30b-99f866651c66)

Should I also update my initial post to make it better describe current version?

GlyphTheWolf commented:
Ok, so I tested it a bit more today and seems that it doesn't work properly with baked meshes. It worked in my tests, because I used 2 `GridMap`s, but if checked baked meshes with single `GridMap` it doesn't work as I would expect. Probably I need to analyze that mesh baking longer.

GlyphTheWolf commented:
Now I think I fixed properly render layers for baked meshes.
Baked meshes are now separated based on octant and other mesh render properties which require separate mesh instance (for now only render layers, but it could be expanded to support also different properties like shadow casting settings etc.).
If user keep same render layers for all `MeshLibrary` items there should be only one baked mesh per octant (same as before), but in case there are different layer settings in one octant then more baked meshes may be generated.
Now render layers works per `MeshLibrary` item even for baked meshes on single `GridMap`:

![RenderLayersMeshLibLightmap](https://github.com/user-attachments/assets/3988a20d-b488-49c0-9a43-15b56734d354)

GlyphTheWolf commented:
Ah, I haven't checked that before, but it seems it is not the first attempt on adding render layers for `GridMap`.
My change seems to be very similar with #75328 and #86992.
Also it is related with #40245.

I think only one solution need to be selected, but whichever it will be, I hope this enhancement will find its way into official release.

BastiaanOlij commented:
Just found this PR, we ran into this limitation as well.

At first glance, looks like nice work @GlyphTheWolf .
Having a quick look at the other two PRs you mentioned,:
- #75328 is indeed very similar to yours but lacks the extra logic you added to make sure the octants are taking into account
- #86992 seems to be more akin to your earlier attempt of just adding a visual layer on the gridmap

Malcolmnixon commented:
This crashes when load an old scene containing a GridMap made befor this patch. Specifically the format of the `baked_meshes` has changed. If I manually edit the tscn file and delete the field (and associated ArrayMesh sub-resource) then I can open the scene and everything works perfectly.

Perhaps the `GridMap::_set` handler could do a sanity-check on the "baked_meshes" p_value and discard if using the old format resulting in a rebake.

AThousandShips commented:
You used a merge commit to update your branch, please use rebase in the future instead, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

clayjohn commented on modules/gridmap/grid_map.cpp:
```diff
@@ -115,7 +120,10 @@ bool GridMap::_get(const StringName &p_name, Variant &r_ret) const {
 		ret.resize(baked_meshes.size());
 		for (int i = 0; i < baked_meshes.size(); i++) {
-			ret[i] = baked_meshes[i].mesh;
+			Array pair;
+			pair.append(baked_meshes[i].mesh);
+			pair.append(baked_meshes[i].item);
+			ret[i] = pair;
```
The problem pointed out by Malcom comes from doing this. You are changing the interpretation of the "baked_meshes" property in the .tscn file.

If you add more data, you need to ensure compatibility. You can do that in two ways:
1. Add the data without changing how old properties are interpreted (i.e. add a new property)
2. In the `_set()` function check the data type that you read, and then gracefully handle upgrading old scenes. I.e. check if `meshes[0]` is an Array or a Mesh. If its a mesh, use the old codepath, if its an array then do the new codepath.

GlyphTheWolf replied:
Yes and I was expecting that old scene using baked meshes won't work after my change.

I am working now on version which use approach 1. (new property instead of updating old one), but I still need to do some testing and find the best way to fix that.

GlyphTheWolf commented:
> You used a merge commit to update your branch, please use rebase in the future instead, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

Ah, right, sorry. I am not used to git and I don't know all that management stuff and keeping history clean. Sometimes with git it's hard for me to even do what I really want.

If that request would be to messy, I could create separate one (after final change will be decided) with all changes put into one commit.

AThousandShips commented:
After finishing up you will need to [squash](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), but to remove the merges you can just rebase and the merge commits should be removed

GlyphTheWolf commented:
> After finishing up you will need to [squash](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), but to remove the merges you can just rebase and the merge commits should be removed

I see. I just did some experiments with that and it seems to work. Now there are only 2 commits in history. For final solution it would be good to squash all of them into one, but for now just for test I wanted to keep only 2 major commits (first with original change and second one with render layers moved to `MeshLibrary`).

GlyphTheWolf commented:
I updated code and now it should work without crashes between versions (both ways old <-> new).

If loading old version of baked meshes to new one, new features just won't be available until `GridMap` baked meshes will be invalidated (for example by modifying `GridMap` cells) and baked again with new version.

If loading new version to old one it should work the same, but new features will be lost (everything on single render layer).

@Malcolmnixon You can check if that new version works with your scene properly. I tested it on my simple test scene and seems to be working fine.

Malcolmnixon commented:
> @Malcolmnixon You can check if that new version works with your scene properly. I tested it on my simple test scene and seems to be working fine.

This is working well for me now - both in loading a Godot 4.3 GridMap scene in a Godot with this patch, and going the other way. You may want to change the title of this PR to "Preserve visual layer of meshes in GridMap nodes".

GlyphTheWolf commented:
I updated title, description in original post and squashed commits into one.

GlyphTheWolf commented:
Hm, interesting. I checked it locally after rebase of my local branch.

For me it looks like baked meshes generation is fine and it is working as expected, because for example when adding decal with specific cull mask it is properly culled before and after baking (so multimeshes and baked meshes keep valid layer masks), but light baking indeed always take into account all layers(?) Not sure why is that, since I am not too familiar with light baking process. Shouldn't layer mask set for instances to rendering server take effect during baking process? Or they need to be passed somehow separately?

> It works well when editing MeshLibrary visibility layers on the Platformer 3D demo, but fails on https://github.com/Calinou/escape-space-3d even after removing the LightmapGI node.

What do you mean by "fails". You mean even cull mask doesn't work properly (like for decals)? In my case (I also tested on project you mentioned) layers works fine for both not baked and baked meshes, only light baking is not working (doesn't respect layer for some reason). I wanted to update this pull request, but I totally messed up my local git branch, so I need to fix it or rebase again before I will be able to push, so I need more time for that.

SpruceBruce64 commented:
I tested the branch with the escape space 3d project and the visual instance layer wasnt doing anything until you tried placing from the gridmap tab and then it worked well (I assume this is when the initial conversion happens). When closing the editor I noticed a bunch of errors that said `ERROR: Index p_idx = 0 is out of bounds (baked_meshes.size() = 0).
   at: get_bake_mesh_instance (modules/gridmap/grid_map.cpp:1518)` but I was also getting those when closing 4.7.1-stable so I assume its unrelated to this change. Just mentioning these in case either is the failure that Calinou was mentioning because otherwise it seemed to work well for me. It also works in my personal project.

There are also merge conflicts with master so it needs to be rebased again.

smix8 commented:
Most of it looks fine but we are not going to modify the main `OctantKey` and related core functions for an optional niche feature. This needs to be solved with its own separated and optional data, e.g. a Hashmap<OctantKey, RandomFeature> lookup.
